cupyx/scipy/interpolate/_interpolate.py

```diff
@@ -0,0 +1,990 @@
+        if(!extrapolate) {
+            out[idx] = -1;
+        } else if(le_or_ge(xp, a, asc)) {
+            out[idx] = 0;
+        } else if(ge_or_le(xp, b, asc)) {
+            out[idx] = total_breakpoints - 2;
+        }
+        return;
+    } else if (xp == b) {
+        out[idx] = total_breakpoints - 2;
+        return;
+    }
+
+    int left = 0;
+    int right = total_breakpoints - 2;
+    int mid;
+
+    if(le_or_ge(xp, *&breakpoints[left + 1], asc)) {
+        right = left;
+    }
+
+    bool found = false;
+
+    while(left < right && !found) {
+        mid = ((right + left) / 2);
+        if(le_or_ge(xp, *&breakpoints[mid], asc)) {
+            right = mid;
+        } else if (geq_or_leq(xp, *&breakpoints[mid + 1], asc)) {
+            left = mid + 1;
+        } else {
+            found = true;
+            left = mid;
+        }
+    }
+
+    out[idx] = left;
+}
+}
+'''
+
+INTERVAL_MODULE = cupy.RawModule(
+    code=INTERVAL_KERNEL, options=('-std=c++11',),)
+
+PPOLY_KERNEL = r"""
+#include <cuda_runtime.h>
+#include <device_launch_parameters.h>
+
+#include <cupy/complex.cuh>
+#include <cupy/math_constants.h>
+
+template<typename T>
+__device__ T eval_poly_1(
+        const double s, const T* coef, long long ci, int cj, int dx,
+        const long long* c_dims, const long long stride_0,
+        const long long stride_1) {
+    int kp, k;
+    T res, z;
+    double prefactor;
+
+    res = 0.0;
+    z = 1.0;
+
+    if(dx < 0) {
+        for(int i = 0; i < -dx; i++) {
+            z *= s;
+        }
+    }
+
+    int c_dim_0 = (int) *&c_dims[0];
+
+    for(kp = 0; kp < c_dim_0; kp++) {
+        if(dx == 0) {
+            prefactor = 1.0;
+        } else if(dx > 0) {
+            if(kp < dx) {
+                continue;
+            } else {
+                prefactor = 1.0;
+                for(k = kp; k > kp - dx; k--) {
+                    prefactor *= k;
+                }
+            }
+        } else {
+            prefactor = 1.0;
+            for(k = kp; k < kp - dx; k++) {
+                prefactor /= k + 1;
+            }
+        }
+
+        int off = stride_0 * (c_dim_0 - kp - 1) + stride_1 * ci + cj;
+        T cur_coef = *&coef[off];
+        res += cur_coef * z * ((T) prefactor);
+
+        if((kp < c_dim_0 - 1) && kp >= dx) {
+            z *= s;
+        }
+
+    }
+
+    return res;
+
+}
+
+template<typename T>
+__global__ void eval_ppoly(
+        const T* coef, const double* breakpoints, const double* x,
+        const long long* intervals, int dx, const long long* c_dims,
+        const long long* c_strides, int num_x, T* out) {
+
+    int idx = blockDim.x * blockIdx.x + threadIdx.x;
+
+    if(idx >= num_x) {
+        return;
+    }
+
+    double xp = *&x[idx];
+    long long interval = *&intervals[idx];
+    double breakpoint = *&breakpoints[interval];
+
+    const int num_c = *&c_dims[2];
+    const long long stride_0 = *&c_strides[0];
+    const long long stride_1 = *&c_strides[1];
```

Comment:
@takagi, I noticed that sometimes some `const` pointers mutate even though they are only accessed and not modified, this has led us to define "copies" by referencing and dereferencing pointer accesses, do you think this is a correct pattern? Or are we doing something wrong when calling the kernels?

cc @ev-br

Reply:
Do you have the compilation failed if you just pass c_strides to `eval_poly_1`?

Reply:
Not a compilation error, but a runtime failure where the first value of `c_strides` changes to a random value.

Reply:
It is supposed that it's a kind of synchronization problem and the memory area is unintentionally broken. Copying the value to register files is not the solution. Let me see it further.

Reply:
I wonder if gh-7240 is related.

Reply:
I checked it on my local. I passed `c_strides[0]` and `c_strides[1]` instead of `stride_0` and `stride_1` to `eval_poly_1()`, but I didn't see the behavior and the tests passed (on CUDA 11.8).

Reply:
> I wonder if https://github.com/cupy/cupy/issues/7240 is related.

We suspect that the cause of gh-7240 is that scipy returns zero or nan depending on its version (right?). So I guess it is not related here.

Reply:
Now I'm not confident with my [comment](https://github.com/cupy/cupy/pull/7204#discussion_r1050275618). Let us leave the lines as far as the CI passes, and revisit them when we have any problems.